Migrate to TypeScript

whymarrh opened the pull request:
This PR migrates the codebase to use TypeScript. This PR also replaces the Babel tooling with tooling from ethereumjs-config, resulting a project structure similar to the account, rlp, etc. libraries. I've made heavy use of `any` as a starting point for future restricting of the types (which will likely have to be accompanied by further refactoring). I've also dropped `safe-buffer` (as outlined in the organizations technical documentation).

holgerd77 commented:
👍

holgerd77 commented on package.json:
```diff
@@ -8,11 +8,17 @@
+    "tsc": "tsc --noEmit",
+    "lint": "npm run format && npm run tslint && npm run tsc",
+    "lint:fix": "npm run format:fix && npm run tslint:fix && npm run tsc",
+    "test": "ts-node node_modules/tape/bin/tape ./test/index.ts"
```
We have now introduced these unified build scripts from ``ethereumjs-config``, see e.g. [in the account library](https://github.com/ethereumjs/ethereumjs-account/blob/master/package.json), please use these here.

whymarrh replied:
https://github.com/ethereumjs/ethereumjs-account/blob/10b46d6be6c87ff07d964e141f290eaa2c2c0f4a/package.json#L27 shows the same script? Which script should I use here?

whymarrh replied:
Ah, yes, the _other_ script entries. I'll update them!

holgerd77 replied:
👍

holgerd77 commented:
Can you also add the return types to the libraries?

holgerd77 commented:
(corrected the comment above, please read on GitHub directly)

holgerd77 commented:
Ok, that's REALLY a lot of any 😄, is it not possible to be more precise in many cases?

holgerd77 commented:
(but if you think that's too much work load on this iteration - so be it - then we do it next round separately)

whymarrh commented:
I am looking at being more specific with the types as I can, but there I think I might have to do a separate PR.

whymarrh commented:
> is it not possible to be more precise in many cases?

The main sticking point is that there are a good few functions in the index file that manipulate the arguments/overload the functions and to express that properly I would need to rewrite the functions themselves and I don't want them to get lost in the diff noise.

holgerd77 commented:
Ah ok, no, them leave as-is, we can also add the return types later.

holgerd77 commented on .editorconfig:
```diff
@@ -0,0 +1,10 @@
+insert_final_newline = true
+
+[*.md]
+indent_size = 4
```
What kind of editor configuration is this here? Even if this might be useful I would prefer not to side introduce here and eventually discuss along a separate PR, can you please remove?

whymarrh replied:
Yup, I'll add this in a separate PR

holgerd77 commented:
Generally: just realizing the scope of this work by scrolling completely through for the first time, really great! 😀 👍

holgerd77 commented:
Can you do one integration test on this and open a PR on the VM pointing the blockchain dependency towards this branch so that we can see if CI runs through? This will largely help build up confidence that this has no side effects (especially after the experiences with the other TS transitions 😆).

whymarrh commented:
> Can you do one integration test on this and open a PR on the VM pointing the blockchain dependency towards this branch so that we can see if CI runs through?

For sure, I'll add that

whymarrh commented:
> Can you do one integration test on this and open a PR on the VM pointing the blockchain dependency towards this branch so that we can see if CI runs through?

Using this branch is a bit tricky with the new build process. The published versions will have compiled TS code where the branch doesn't. I could add a temporary step to ethereumjs/ethereumjs-vm#469 that compiles the source, how does that sound?

holgerd77 commented:
@whymarrh Whatever is easier for you, you can also do a local test run and post the results here.

holgerd77 commented:
This lacks a ``prettier.config.js`` file, see e.g. [the one from the account library](https://github.com/ethereumjs/ethereumjs-account/blob/master/prettier.config.js).

holgerd77 commented on src/cache.ts:
```diff
@@ -0,0 +1,33 @@
+    }
+    this._cache.del(key);
+  }
+}
```
Ok.

holgerd77 commented on src/util.ts:
```diff
@@ -0,0 +1,62 @@
+  bodyKey,
+  numberToHashKey,
+  hashToNumberKey
+};
```
Why is there a type annotation only for ``bufBE8``? I generally find this ``bufBE8`` line hard to read for a non-TypeScript pro.

whymarrh replied:
I don't understand, which line are you referring to?

holgerd77 replied:
``const bufBE8: (n: BN) => Buffer = (n: BN) => n.toArrayLike(Buffer, "be", 8);`` (the ``: (n: BN) => Buffer`` is a type annotation if I am not mistaken?) vs ``const tdKey = (n: BN, hash: Buffer) =>
  Buffer.concat([headerPrefix, bufBE8(n), hash, tdSuffix]);`` and the rest.

whymarrh replied:
I still don't understand the question. What part of the type annotation is weird?

whymarrh replied:
Updated to make it a bit less verbose

holgerd77 commented on src/dbManager.ts:
```diff
@@ -0,0 +1,178 @@
+  batch(ops: Array<any>): Promise<any> {
+    return this._db.batch(ops);
+  }
+}
```
Ok.

holgerd77 commented:
Also [.prettierignore](https://github.com/ethereumjs/ethereumjs-account/blob/master/.prettierignore).

holgerd77 commented on package.json:
```diff
@@ -38,19 +44,30 @@
+    "typedoc": "^0.14.2",
+    "typedoc-plugin-markdown": "^1.1.27",
+    "typescript": "^3.3.3333",
+    "typestrict": "^1.0.2"
```
``types`` directive missing (below ``main``)

whymarrh replied:
Added

whymarrh commented:
I've got ethereumjs/ethereumjs-vm#469 to build successfully based on (a form of) these changes

holgerd77 commented:
This has now linting errors since ``prettier`` is activated and CI is failing.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+  _initLock: any;
+  _putSemaphore: any;
+  _staleHeadBlock: any;
+  _staleHeads: any;
```
Nice resorting and - generally - property collection, finally gives some overview what's inside. 😄  👍

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+      }
+      this._initLock.go();
+    });
+  }
```
``constructor()``: Ok.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+        }
+        cb();
+      }
+    );
```
``get meta()``, ``_init()``: Ok.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+        self._headBlock = genesisHash;
+      }
+    }
+  }
```
``getHeads()``: Ok.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+      },
+      cb
+    );
+  }
```
Ok till here.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+        this._putBlockOrHeader(block, done, isGenesis);
+      }, cb);
+    });
+  }
```
``lockUnlock`` extracted, good.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+        this._putBlockOrHeader(header, done);
+      }, cb);
+    });
+  }
```
Both ok.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+        next();
+      }
+    }
+  }
```
``_putBlockOrHeader()``: Ok.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+    }
+
+    nextBlock(blockId);
+  }
```
Ok until here.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+   * Gets block details by its hash
+   * @deprecated
+   */
+  getDetails(_: string, cb: any) {
```
Is this ``hash`` -> ``_`` replacement some convention to mark things as deprecated? Other reason?

whymarrh replied:
The `hash` argument isn't used so it's a convention to note that it's not used

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+
+  _saveHeads(cb: any) {
+    this._batchDbOps(this._saveHeadOps(), cb);
+  }
```
Ok till here.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+        cb();
+      }
+    });
+  }
```
Ok till here.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+        cb
+      );
+    });
+  }
```
Ok till here.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+    function run(cb2: any) {
+      let block: any;
+
+      async.series([getBlock, runFunc], function(err: any) {
```
Is this ``err`` here non-optional?

whymarrh replied:
It'll be `null` or a `error`, I do believe

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+        func(block, reorg, cb3);
+      }
+    }
+  }
```
Ok.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+      }
+    );
+  }
+
```
Removed unused internal ``_getBody()`` function, ok.

holgerd77 commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+      }
+    });
+  }
+}
```
Everything ok, looks good.

holgerd77 commented on test/index.ts:
```diff
@@ -0,0 +1,872 @@
+      cb(err, db, genesis);
+    }
+  );
+}
```
Ok.

whymarrh commented on src/index.ts:
```diff
@@ -0,0 +1,1003 @@
+    });
+  }
+
+  _delBlock(blockHash: any, cb: any) {
```
I'll update this to use `Buffer`

whymarrh replied:
Good catch

whymarrh commented:
@holgerd77 updated again, let me know how it looks now

holgerd77 commented:
Hi @whymarrh, thanks for the updates, can you please have a final look at the CI run some time after you push and check that this is ok? Otherwise this is always blocking.

Linting is still failing, probably ``prettier`` fix run still needs to be applied.

holgerd77 commented:
@whymarrh Can you do this final ``prettier`` run so that we can get this over the finish line?

whymarrh commented:
Updated. We'll also want to drop 83fb9c0 from this PR or before we publish.

holgerd77 commented:
@whymarrh Not completely getting it. Why did you introduce this in the first place and why didn't you remove now along this update? Is this something which needs further work?

whymarrh commented:
@holgerd77 updated and removed 83fb9c0. Apologies for the confusion, what wasn't clear? To be able to use this package via Git in https://github.com/ethereumjs/ethereumjs-vm/pull/469 the `files` entry needed to be dropped. I've dropped them now and we can deal with that PR separately now—that PR no longer works but that's fine, once we publish a version of this package I can update that PR with the version from the registry.

From that PR's description:

> Note: while [[ethereumjs/ethereumjs-blockchain#92](https://github.com/ethereumjs/ethereumjs-blockchain/pull/92)] is in-progress this will be pointing to a Git URL of my fork. Once the PR is merged (and the tests here will serve as a useful smoke test) this PR will be updated with the correct version on the registry.
